public/app/features/explore/hooks/useStateSync.ts

```diff
@@ -0,0 +1,409 @@
+      if (Object.values(urlPanes).some(({ range }, _, [{ range: firstRange }]) => !isEqual(range, firstRange))) {
+        dispatch(syncTimesAction({ syncedTimes: false }));
+      }
+
+      for (const [exploreId, urlPane] of Object.entries(urlPanes) as Array<[ExploreId, ExploreUrlState]>) {
+        const { datasource, queries, range, panelsState } = urlPane;
+
+        const statePane = statePanes[exploreId];
+
+        if (statePane !== undefined) {
+          const update = urlDiff(urlPane, getUrlStateFromPaneState(statePane));
+
+          Promise.resolve()
+            .then(async () => {
+              if (update.datasource) {
+                await dispatch(changeDatasource(exploreId, datasource));
+              }
+              return;
+            })
+            .then(() => {
+              if (update.range) {
+                dispatch(updateTime({ exploreId, rawRange: range }));
+              }
+
+              if (update.queries) {
+                dispatch(setQueriesAction({ exploreId, queries: withUniqueRefIds(queries) }));
+              }
+
+              if (update.queries || update.range) {
+                dispatch(runQueries({ exploreId }));
+              }
+            });
+        } else {
+          // This happens when browser history is used to navigate.
+          // In this case we want to initialize the pane with the data from the URL
+          // if it's not present in the store. This may happen if the user has navigated
+          // from split view to non-split view and then back to split view.
+          dispatch(
+            initializeExplore({
+              exploreId,
+              datasource,
+              queries: withUniqueRefIds(queries),
+              range,
+              panelsState,
+            })
+          );
+        }
+      }
+
+      // Close all the panes that are not in the URL but are still in the store
+      // ie. because the user has navigated back after opening the split view.
+      Object.keys(statePanes)
+        .filter((keyInStore) => !Object.keys(urlPanes).includes(keyInStore))
+        .forEach((paneId) => dispatch(splitClose(paneId as ExploreId)));
+    }
+
+    // This happens when the user first navigates to explore.
+    // Here we want to initialize each pane initial data, wether it comes
+    // from the url or as a result of migrations.
+    if (!isURLOutOfSync && initState.current === 'notstarted') {
+      initState.current = 'pending';
+
+      // Clear all the panes in the store first to avoid stale data.
+      dispatch(clearPanes());
+
+      Promise.all(
+        Object.entries(urlPanes).map(([exploreId, { datasource, queries, range, panelsState }]) => {
+          return getPaneDatasource(datasource, queries, orgId, !!exploreMixedDatasource).then(
+            async (paneDatasource) => {
+              return Promise.resolve(
+                // FIXME: In theory, given the Grafana datasosurce will always be present, this should always be defined.
```

Comment:
```suggestion
                // FIXME: In theory, given the Grafana datasource will always be present, this should always be defined.
```

Do we need `FIXME` here?

Reply:
good point. removed it in cc11aee646